linalg/vector.cpp

```diff
@@ -28,6 +28,209 @@
 namespace mfem
 {
 
+#if defined(MFEM_USE_CUDA) or defined(MFEM_USE_HIP)
+/**
+ * Reducer for helping to compute L2-norms. Given two partial results:
+ * a0 = sum_i (|v_i|/a1)^2
+ * b0 = sum_j (|v_j|/b1)^2 (j disjoint from i for vector v)
+ * computes:
+ * a1 = max(a1, b1)
+ * a0 = (a1 == 0 ? 0 : sum_{k in union(i,j)} (|v_k|/a1)^2)
+ *
+ * This form is resiliant against overflow/underflow, similar to std::hypot
+ */
+struct L2Reducer
+{
+   using value_type = DevicePair<real_t, real_t>;
+   MFEM_HOST_DEVICE void join(value_type& a, const value_type &b) const
+   {
+      real_t scale = fmax(a.second, b.second);
+      if (scale > 0)
+      {
+         real_t s = a.second / scale;
+         a.first *= s * s;
+         s = b.second / scale;
+         a.first += b.first * s * s;
+         a.second = scale;
+      }
+   }
+
+   MFEM_HOST_DEVICE void init_val(value_type &a) const
+   {
+      a.first = 0;
+      a.second = 0;
+   }
+};
+
+/**
+ * Reducer for helping to compute Lp-norms. Given two partial results:
+ * a0 = sum_i (|v_i|/a1)^p
+ * b0 = sum_j (|v_j|/b1)^p (j disjoint from i for vector v)
+ * computes:
+ * a1 = max(a1, b1)
+ * a0 = (a1 == 0 ? 0 : sum_{k in union(i,j)} (|v_k|/a1)^p)
+ *
+ * This form is resiliant against overflow/underflow, similar to std::hypot
+ */
+struct LpReducer
+{
+   real_t p;
+   using value_type = DevicePair<real_t, real_t>;
+   MFEM_HOST_DEVICE void join(value_type& a, const value_type &b) const
+   {
+      real_t scale = fmax(a.second, b.second);
+      if (scale > 0)
+      {
+         a.first = a.first * pow(a.second / scale, p) +
+                   b.first * pow(b.second / scale, p);
+         a.second = scale;
+      }
+   }
+
+   MFEM_HOST_DEVICE void init_val(value_type &a) const
+   {
+      a.first = 0;
+      a.second = 0;
+   }
+};
+
+static Array<real_t>& vector_workspace()
+{
+   static Array<real_t> instance;
+   return instance;
+}
+
+static Array<DevicePair<real_t, real_t>> &Lpvector_workspace()
+{
+   static Array<DevicePair<real_t, real_t>> instance;
+   return instance;
+}
+
+static real_t devVectorMin(int size, const real_t *m_data)
+{
+   real_t res = infinity();
+   reduce(
+      size, res,
+   [=] MFEM_HOST_DEVICE(int i, real_t &r) { r = fmin(r, m_data[i]); },
+   MinReducer<real_t> {}, true, vector_workspace());
+   return res;
+}
+
+static real_t devVectorMax(int size, const real_t *m_data)
```

Comment:
Maybe it would be cleaner if this was a lambda inside `Vector::Max`. It seems like the only reason to factor it out as a function is because it's called in both the CUDA and HIP cases.

Maybe we want a generic "using accelerator" check, and if that's true, call the "dev" version, rather than separate CUDA and HIP checks.